### DISCUSSION TITLE
Suggestion: swift-html-vapor-provider & swift-html-vapor-example

### DISCUSSION THREAD

**Discussion by marc-medley** [pointfreeco/swift-html-vapor#10] [Ideas]
For `swift-html` and Vapor, it would be helpful if two separate repositories existed: 

* `swift-html-vapor-provider`: package which provides `swift-html` components a Vapor project which support Vapor's use of SwiftNIO and Services. 

    * [SwifthtmlProvider](https://github.com/VaporExamplesLab/eval-vapor-leaf-stencil/blob/master/Sources/App/Services/SwifthtmlProvider.swift) used to register `swift-html` as a Vapor Service.
    * [SwifthtmlRenderer](https://github.com/VaporExamplesLab/eval-vapor-leaf-stencil/blob/master/Sources/App/Services/SwifthtmlRenderer.swift) asynchronous Response rendering.

* `swift-html-vapor-example`: Vapor example which uses the `swift-html-vapor-provider`

    * [SwifthtmlController](https://github.com/VaporExamplesLab/eval-vapor-leaf-stencil/blob/master/Sources/App/Controllers/SwifthtmlController.swift) sets up routes and where to locate additional non-public resources.
    * [config](https://github.com/VaporExamplesLab/eval-vapor-leaf-stencil/blob/master/Sources/App/configure.swift) `services.register(SwiftOrganicProvider())`
    * [SwifthtmlTemplates](https://github.com/VaporExamplesLab/eval-vapor-leaf-stencil/tree/master/Sources/App/SwifthtmlTemplates) example template layouts :-)
  
[`VaporExamplesLab/eval-vapor-leaf-stencil`](https://github.com/VaporExamplesLab/eval-vapor-leaf-stencil) is a hands-on look at pure Swift templating language, `swift-html` and just plain organic Swift.  You are welcome to leverage any of the [eval-vapor-leaf-stencil](https://github.com/VaporExamplesLab/eval-vapor-leaf-stencil). Feedback is also welcome.

**Comment by mbrandonw** [comment]
@marc-medley Hey Marc, these are great suggestions! I'm not sure if two separate repos are necessary though, this one repo can house both the core library and the sample application. That is actually how it works now, you can run `swift run HtmlVaporSupportExample` in this repo and it will fire up a sample Vapor application that is rendered using the swift-html library.

I definitely agree that we should beef up the sample app a lot more to make it clearer how to use. I've taken a first pass at that in this PR #6, which shows how to handle multiple routes, break views into separate files, and provide a layout that is responsible for rendering header, footer, etc. We will follow up soon to also implement the `Provider` and `Service` protocols, and to beef up the sample app even more to use a controller and parse parameters from the URL.

Thanks for the nudge to get this done, and we'll write back if we have any questions about your Vapor sample code!

**Comment by mbrandonw** [comment]
After looking over these examples it seems like these protocols (`Service` and `Provider`) provide functionality for templating languages that swift-html doesn't need, since you can write plain Swift and pass values accordingly. We might be overlooking something simple, so maybe you could provide a more concrete example of how swift-html may best utilize these protocols?

**Comment by marc-medley** [comment]
Regardless of whether a second `swift-html-vapor-provider` is created or not, some information of how swift-html can work with dynamic conent in both examples and documentation would helpful.

Here are some thoughts for you consideration:

**Dynamic Content Latency**

An http Response with dynamic content may need to be composed from parts with varing latency. `swift-html` would be low latency. File, database queries, and secondary http urls would be higher latency.

> swift-html doesn't need, since you can write plain Swift and pass values accordingly

So, what's potentially missing in the "swift-html-vapor" example is more than just a `swift-html` layout per se. A pattern for how low latency `swift-html` content would combine with other higher latency content components is also useful.

**Real World Template Layouts**

Since `swift-html` would not use files like templating language, then it would be reasonable that each swift-html template would be a `struct` or `class`.

One approach could be:

``` swift 
struct SwifthtmlTemplateDynamicContent: SwifthtmlTemplateProtocol {
    public func render<E>(context: E) -> Html.Node where E: Encodable  {
         guard let dictionary = context as? [String:String],
            let urlStr = dictionary["urlKey"],
            let filepathStr = dictionary["filepathKey"],
            let sqlqueryStr = dictionary["sqlqueryKey"]
            else { /* */ }

        let urlContent = getContent(urlStr)
        let fileContent = getContent(urlStr)
        let queryContent = getContent(urlStr)

        let document: Node = [
            .doctype,
            .html(  
                /* use both siwft-html and higher latency content pieces here */
            )
        ]
        return document
    }
}

```

Another approach could be to retrieve higher latency content first. Separately.  However, a _content template with mixed latency_ can be handled within the Vapor Services…

**Vapor Services Provider Pattern**

Vapor 3 (unlike prior versions) is asynchronous and built on SwiftNIO `EventLoopPromise` (alias `Promise`) and `EventLoopFuture` (alias `Future`). This is where the Vapor Services Provider/Renderer pattern is well suited for a template which has a _combination of lower and higher latency content._

``` swift
public final class SwifthtmlRenderer: Service {
    // …    
    
    init(using container: Container) throws {
        let directoryConfig = try container.make(DirectoryConfig.self)
        // …
        self.container = container
    }

    public func render<E>(template: SwifthtmlTemplate, _ context: E)
        -> Future<Html.Node> where E : Encodable {
            let promiseNode: Promise<Html.Node> = container.eventLoop.newPromise(Html.Node.self)
            
            DispatchQueue.global().async {
                let node = template.render(context: context)
                return promiseNode.succeed(result: node)
            }
            let futureNode = promiseNode.futureResult
            return futureNode // Future<Html.Node>
    }
}
```

The Vapor Service Provider allows a clean separation of the layout of content and the mechanism a asynchronous deliverly.  And, can allow content layout to focus on _all content elements in one place_.

**GitHub "topic:__" Search Pattern**

_Conceptually, a `swift-html-vapor-provider` would be the turnkey Vapor Service factory for a generic swift-html Renderer which Vapor projects (such as `swift-html-vapor`) could use.  This generic  Renderer would allow a Vapor web developer to more quickly use swift-html with other site content (both static and dynamic)._

_If such a `swift-html-vapor-provider` (or `vapor-swift-html-provider`) existed, it could found with other Vapor Providers on the GitHub search [topic:vapor-provider ⇗](https://github.com/search?q=topic%3Avapor-provider)_

**Comment by stephencelis** [comment]
Hey @marc-medley! I think I'm still failing to see what this abstraction brings to the table for a solution like swift-html. I've pushed a commit that adds a coupla dynamic examples: https://github.com/pointfreeco/swift-html-vapor/commit/96751e35ecf378783449b5dcb4e48595629b7a45

First I've updated the `ResponseEncodable` implementation to be asynchronous. swift-html rendering should be relatively quick, so I'm not sure the thread hop is worth it, but it's possible to bake this into the library:

https://github.com/pointfreeco/swift-html-vapor/blob/96751e35ecf378783449b5dcb4e48595629b7a45/Sources/HtmlVaporSupport/HtmlVaporSupport.swift#L6-L8

We'll consider whether it's worth doing so! For now it's easy enough for the end user of the library to go async for larger renderings.

Secondly, here's how we typically add dynamicism to an endpoint: we use a function.

https://github.com/pointfreeco/swift-html-vapor/blob/96751e35ecf378783449b5dcb4e48595629b7a45/Sources/HtmlVaporSupportExample/Add.swift#L3-L8

This is a free function in the module but folks can embed their function as statics in type namespaces should they want to. It takes the data the view needs and returns a `Node` that uses this data.

Then the router can freely pass the parameters to this function:

https://github.com/pointfreeco/swift-html-vapor/blob/96751e35ecf378783449b5dcb4e48595629b7a45/Sources/HtmlVaporSupportExample/main.swift#L23-L31

The one last thing you mentioned:

> An http Response with dynamic content may need to be composed from parts with varing latency. swift-html would be low latency. File, database queries, and secondary http urls would be higher latency.

I've simulated IO with a sleep/delay here:

https://github.com/pointfreeco/swift-html-vapor/blob/96751e35ecf378783449b5dcb4e48595629b7a45/Sources/HtmlVaporSupportExample/main.swift#L33-L40

While I've manually created a promise and fulfilled it with GCD, this is just to show that you could use any kind of library that vends `EventLoopFuture`s and, once you have all your data prepared, the final `EventLoopFuture` needs to return the `Node` to be rendered.

If there's anything missing from the "provider" approach, please let me know! I can see that using protocols is an attractive way of building up applications that rely heavily on those conformance hooks, but swift-html is so lightweight that I'm not really seeing what those extra layers of abstraction allows. What seems to be another key benefit of swift-html over other templating systems is that you don't need to build up those extra types, conformances, etc.

**Comment by marc-medley** [comment]
@stephencelis Thanks for the updated examples. 

I'm not so much trying to create any _new_ abstractions approaches. Rather it's an exploration of  _if and/or how_ swift-html fits into the _existing_ abstraction approaches of the Vapor architecture. In this case, the [Vapor Services](https://docs.vapor.codes/3.0/service/services/) where Vapor View rendering archetype exists.

I look forward to exploring this topic more, and providing subsequent feedback/comment. ... once  swift-html with dependencies can build on a Swift 4.2 system.

This is definitely a path finding exercise complete with some unexpected detours.
